src/AllenAlgebra.elm

```diff
@@ -0,0 +1,72 @@
+module AllenAlgebra exposing (..)
+
+{-| The relations defined by an allen algebra.
+
+<https://www.wikiwand.com/en/Allen's_interval_algebra#/Relations>
+
+-}
+
+
+type AllenRelation
+    = Overlaps
+    | OverlapsInverse
+    | During
+    | DuringInverse
+    | Meets
+    | MeetsInverse
+    | Starts
+    | StartsInverse
+    | Finishes
+    | FinishesInverse
+    | Before
+    | After
+    | Equal
+
+
+relation : { r | start : Int, end : Int } -> { p | start : Int, end : Int } -> AllenRelation
+relation range1 range2 =
+    case
+        ( compare range1.start range2.start
+        , compare range1.start range2.end
+        , compare range1.end range2.start
+        , compare range1.end range2.end
+        )
+    of
+        ( _, _, LT, _ ) ->
```

Comment:
🤔 This feels a bit off to me, but I'm unsure how to word it out. Forgive the rambling.

It looks like there's an implicit invariant that `start <= end`. We can do a few things about that:
* Document the invariant. Mentioning that if you pass in a record where `start > end` is undefined behavior is one way to do that.
* Ignore the invariant. We're kind fo doing that now, so this would amount to doing nothing.
* Remove the invariant. If we handle all 81 cases, we have the invariant under control. That number was calculated by:
    ```
    (Order, Order, Order, Order)
      ~ Order * Order * Order * Order
      ~ 3 * 3 * 3 * 3
      ~ 81
    ```
    We can probably ignore _some_ of the cases. Maybe we can reorder the cases we have now?
* Type the invariant. I don't think this will be so nice in elm.

Am I off-base here?

Reply:
Good point! I managed to remove the invariant by dropping the automatic assumption that `start` is lower than `end`. We now compare both values to find out which of the two is the lower bound and which the upper bound. So if you want to define a range `{ start = 10, end = 5 }`, that's interpreted in the same way as the range `{ start = 5, end = 10 }`.